### ISSUE TITLE
discarded statement and immediate invocation contexts are not mutually exclusive

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#51573]
|  |  |
| --- | --- |
| Bugzilla Link | [52231](https://llvm.org/bz52231) |
| Version | trunk |
| OS | Windows NT |
| CC | @zygoloid |

## Extended Description 
Clang currently rejects this valid C++2b program:
```
consteval int *make() { return new int; }
auto f() {
  if constexpr (false) {
    if consteval {
      // Immediate function context, so call to `make()` is valid.
      // Discarded statement context, so `return 0;` is valid too.
      delete make();
      return 0;
    }
  }
  // FIXME: this error should not happen.
  return 0.0; // expected-error {{'auto' in return type deduced as 'double' here but deduced as 'int' in earlier return statement}}
}

```
https://godbolt.org/z/Y81nf63f7

This seems to happen because we treat a discarded statement evaluation context as being mutually exclusive with an immediate invocation, but as this code shows, they are not mutually exclusive in practice.

See https://reviews.llvm.org/D112089 for details.

**Comment by cor3ntin** [comment]
@AaronBallman I think you can close this https://reviews.llvm.org/D113749

**Comment by AaronBallman** [comment]
This issue was addressed in 6eeda06c1d22da2b9fe96a2569a8a0f8e4f36880 with a follow-up fix in  2334314550724812bb94e36c6b5df7d665bdbd9d.
